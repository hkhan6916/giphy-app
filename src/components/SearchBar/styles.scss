.search-bar {
    width: 100%;
    height: 100%;
  }
  .input-container {
    background-color: #ffffff;
    border: solid 1px #8800ff;
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
    & > input {
      all: unset;
      width: 100%;
      height: 2.5rem;
      color: #000000;
      margin-left: 1rem;
      &::-webkit-search-cancel-button {
        display: none;
      }
      &::placeholder {
        color: #818181;
      }
    }
  
    .close-button {
      all: unset;
      cursor: pointer;
      display: flex;
      justify-content: center;
      align-items: center;
      & .close-icon {
        fill: #8800ff;
        margin-right: 20px;
        width: 2rem;
        height: 2rem;
      }
    }
  }
  .search-error {
    color: #ff0000;
  }
  
  // desktop specific styles
  @media only screen and (min-width: 600px) {
    .search-button {
      display: block;
    }
    .search-button-icon.solid {
      display: none;
    }
  
    .input-container {
      & .search-icon {
        display: block;
      }
    }
  }
  