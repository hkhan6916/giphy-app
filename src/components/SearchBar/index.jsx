import React, { useState } from 'react';
import { AiOutlineCloseCircle } from 'react-icons/ai';
import useTypingDebounce from '../../hooks/useTypingDebounce';

import './styles.scss';

/**
 * SearchBar
 *
 * Renders a custom Search input container for a search bar with custom styles
 *
 * @param {() => void} [onCancel] - The function to call when the close button is clicked
 * @param {() => void} [onTypingEnd] - The function to call when typing stops
 * @param {number} [debounceDelay] - The period of time to wait before calling the onTypingEnd function
 * @param {Ref} [innerRef] - The ref to forward
 */
const SearchBar = ({
    debounceDelay,
    onTypingEnd = () => null,
    name,
    onCancel,
    children,
    className,
    innerRef,
    error,
    style,
    'data-testId': dataTestId,
    ...restOfProps
}) => {
    const [value, setValue] = useState('');
    const [enableDebounce, setEnableDebounce] = useState(false);

    useTypingDebounce({
        onTypingEnd: onTypingEnd,
        value,
        delay: debounceDelay || 100,
        enabled: enableDebounce,
    });

    const handleChange = (value) => {
        // This condition makes sure useTypingDebounce does not run when the component first renders
        if (!enableDebounce && value) setEnableDebounce(true);
        setValue(value);
    };

    const handleCancel = () => {
        // We reset the field and also call the onCancel which could be used for something like clearing the search results externally
        setValue('');
        if (onCancel) {
            onCancel();
        }
    };

    return (
        <div className='search-bar' style={style}>
            <div
                className='input-container'
                style={{
                    display: 'flex',
                    alignItems: 'center',
                    margin: '0 0 2px 0',
                }}>
                <input
                    className='input'
                    alt='Input Box'
                    type='search'
                    name={name}
                    ref={innerRef}
                    value={value}
                    onChange={(e) => handleChange(e.target.value)}
                    {...restOfProps}
                />
                {children}
                {!!value && (
                    <button
                        aria-label='Cancel Button'
                        type='button'
                        className='close-button'
                        onClick={() => handleCancel()}>
                        <AiOutlineCloseCircle
                            title='Close Icon'
                            className='close-icon'
                        />
                    </button>
                )}
            </div>
            {error && (
                <small data-testid='search-error' className='search-error'>
                    {error}
                </small>
            )}
        </div>
    );
};

export default SearchBar;
