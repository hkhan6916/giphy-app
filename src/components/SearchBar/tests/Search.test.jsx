import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import SearchBar from '../index';
import '@testing-library/jest-dom';

expect.extend(toHaveNoViolations);
const { getComputedStyle } = window;
window.getComputedStyle = (elt) => getComputedStyle(elt);

describe('SearchBar', () => {
    test('should render correctly', async () => {
        render(
            <SearchBar
                name='testSearchBar'
                placeholder='test search placeholder'
            />
        );

        const renderedSearchBar = screen.getByPlaceholderText(
            'test search placeholder'
        );
        expect(renderedSearchBar).toMatchSnapshot();
    });
    test('should render error correctly', async () => {
        const { getByText } = render(
            <div>
                <SearchBar
                    name='errorSearch'
                    placeholder='test search placeholder'
                    data-testid='searchTest'
                    error='some error'
                />
                <button type='submit' data-testid='submitButton'>
                    Submit
                </button>
            </div>
        );

        const renderedError = getByText('some error');

        expect(renderedError).toHaveTextContent('some error');
    });
    test('should not have any programatically detectable a11y issues', async () => {
        const { getByTestId } = render(
            <SearchBar
                name='testSearchBar'
                placeholder='test search placeholder'
                data-testid='searchTest'
            />
        );

        expect(await axe(getByTestId('searchTest'))).toHaveNoViolations();
    });
});
