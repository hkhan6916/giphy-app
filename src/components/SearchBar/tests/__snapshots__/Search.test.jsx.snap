// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly 1`] = `
<input
  alt="Input Box"
  class="input"
  name="testSearchBar"
  placeholder="test search placeholder"
  type="search"
  value=""
/>
`;
